### PR TITLE
Simplify the SkiaRenderer C++ constructor

### PR DESCRIPTION
**Pull request by tronical**
Remove the size parameter, and instead allocate the
renderer's surface on demand.

### CODE REVIEW

File: internal/renderers/skia/lib.rs [hunk 1]
```diff
@@ -60,42 +64,48 @@ pub struct SkiaRenderer {
...
 }
 
 impl SkiaRenderer {
     /// Creates a new renderer is associated with the provided window adapter.
     pub fn new(
-        window_handle: raw_window_handle::WindowHandle<'_>,
-        display_handle: raw_window_handle::DisplayHandle<'_>,
-        size: PhysicalWindowSize,
+        window_handle_provider: impl raw_window_handle::HasWindowHandle + 'static,
+        display_handle_provider: impl raw_window_handle::HasDisplayHandle + 'static,
```
Comment (tronical) [inline internal/renderers/skia/lib.rs:76]: I'm not entirely sure about the API here. In practice we end up boxing an Rc here. It would be nice to take an `Rc<dyn ...>`, but that requires some more gymnastics in the winit backend.

File: internal/renderers/skia/lib.rs [hunk 2]
```diff
@@ -60,42 +64,48 @@ pub struct SkiaRenderer {
...
+            surface: Default::default(),
+            window_handle_provider,
+            display_handle_provider,
         })
     }
 
     /// Render the scene in the previously associated window. The size parameter must match the size of the window.
     pub fn render(&self) -> Result<(), i_slint_core::platform::PlatformError> {
+        let Some(surface) = self.surface.get() else {
+            return Err(format!("Skia Renderer: render() called before a resize event was dispatched to the slint::Window").into());
```
Comment (ogoffart) [inline internal/renderers/skia/lib.rs:94]: don't we just ignore this error in the software renderer?

File: internal/renderers/skia/lib.rs [hunk 3]
```diff
@@ -60,42 +64,48 @@ pub struct SkiaRenderer {
...
 }
 
 impl SkiaRenderer {
     /// Creates a new renderer is associated with the provided window adapter.
     pub fn new(
-        window_handle: raw_window_handle::WindowHandle<'_>,
-        display_handle: raw_window_handle::DisplayHandle<'_>,
-        size: PhysicalWindowSize,
+        window_handle_provider: Rc<dyn raw_window_handle::HasWindowHandle>,
+        display_handle_provider: Rc<dyn raw_window_handle::HasDisplayHandle>,
```
Comment (ogoffart) [inline internal/renderers/skia/lib.rs:76]: From an API point of view, this is not ideal.   This always force the handle i a Rc.  which may not be easy.
I think using a `raw_window_handle::WindowHandle<'_>`  is the right API.

Is it worth trying to solve this problem? what's wrong with passing the size?

Could this be done with a `init(handle, handle)`  instead?

File: internal/renderers/skia/lib.rs (hunk 2 above)
Comment (tronical) [reply internal/renderers/skia/lib.rs:94]: Yes. I'd also prefer to ignore it here TBH. If you're also ok with that then I'll change it. But it is kind of a "trap" :)

File: internal/renderers/skia/lib.rs (hunk 3 above)
Comment (tronical) [reply internal/renderers/skia/lib.rs:76]: > Is it worth trying to solve this problem?

That's a good question.

On the one hand it doesn't make sense to initialise a renderer with a given handle to a native window but say: "Hey, the size is not known yet". Of course the native window has a size, it must have - I think.

On the other hand the size can be zero initially, which doesn't work here. And the Rc<> here helps to ensure that the window handle remains valid throughout the lifetime - this really adds safety to the API. And thirdly, our API doesn't allow for the renderer to be created lazily (because of the renderer() getter on the WindowAdapter).

In conclusion, I'll close this PR and we stick to the existing API. If we later find that we can also do this conveniently without the size then we can still add an overload to the C++ SkiaRenderer API.

At least we know that the current approach definitely works for the users we know of and tried this with.

**Comment by tronical** [comment]
Deferred until maybe a later point in time :)
